PermissionScope/Structs.swift

```diff
@@ -6,38 +6,118 @@
+    case reminders
+    #endif
+    #if PermissionScopeRequestEventsEnabled
+    case events
+    #endif
+    #if PermissionScopeRequestBluetoothEnabled
+    case bluetooth
+    #endif
+    #if PermissionScopeRequestMotionEnabled
+    case motion
+    #endif
+
+    var details: PermissionTypeDetails {
+        #if PermissionScopeRequestContactsEnabled
+            if self == .contacts {
+                return PermissionTypeDetails(description: "Contacts", status: PermissionScope().statusContacts())
+            }
+        #endif
+        #if PermissionScopeRequestLocationEnabled
+            if self == .locationAlways {
+                return PermissionTypeDetails(description: "LocationAlways", prettyDescription: "Location", isALocationType: true, status: PermissionScope().statusLocationAlways())
+            }else if self == .locationInUse {
+                return PermissionTypeDetails(description: "LocationInUse", prettyDescription: "Location", isALocationType: true, status: PermissionScope().statusLocationInUse())
+            }
+        #endif
+        #if PermissionScopeRequestNotificationsEnabled
+            if self == .notifications {
+                return PermissionTypeDetails(description: "Notifications", status: PermissionScope().statusNotifications())
+            }
+        #endif
+        #if PermissionScopeRequestMicrophoneEnabled
+            if self == .microphone {
+                return PermissionTypeDetails(description: "Microphone", status: PermissionScope().statusMicrophone())
+            }
+        #endif
+        #if PermissionScopeRequestCameraEnabled
+            if self == .camera {
+                return PermissionTypeDetails(description: "Camera", status: PermissionScope().statusCamera())
+            }
+        #endif
+        #if PermissionScopeRequestPhotoLibraryEnabled
+            if self == .photos {
+                return PermissionTypeDetails(description: "Photos", status: PermissionScope().statusPhotos())
+            }
+        #endif
+        #if PermissionScopeRequestRemindersEnabled
+            if self == .reminders {
+                return PermissionTypeDetails(description: "Reminders", status: PermissionScope().statusReminders())
+            }
+        #endif
+        #if PermissionScopeRequestEventsEnabled
+            if self == .events {
+                return PermissionTypeDetails(description: "Events", status: PermissionScope().statusEvents())
+            }
+        #endif
+        #if PermissionScopeRequestBluetoothEnabled
+            if self == .bluetooth {
+                return PermissionTypeDetails(description: "Bluetooth", status: PermissionScope().statusBluetooth())
+            }
+        #endif
+        #if PermissionScopeRequestMotionEnabled
+            if self == .motion {
+                return PermissionTypeDetails(description: "Motion", status: PermissionScope().statusMotion())
+            }
+        #endif
+        return PermissionTypeDetails()
```

Comment:
I have one issue with the `details` pattern here: Let's say that I want to get the `description` of a permission type, so I call `someType.details.description`. In doing so, I've created an instance of `PermissionTypeDetails`, which immediately invokes the status check for that permission... I don't know all the details about each of these status checks, but I can imagine that some of them might be costly? In fact, in the calling code that cares about status, there is a wrapper function `statusForPermission` with a completion handler built in, as though the library architecture allows for cases where maybe the check happens asynchronously on a background thread.

So I think maybe what I would prefer to see is a way for the status check to be invoked more lazily than this. Perhaps rather than storing a `PermissionStatus`, the details object could instead store a `() -> Bool` closure that captures the work of checking the status. Then you could do something like `someType.details.status()` to check the current value.